### PR TITLE
Rewrite `target` for <a> during [href] mutate

### LINKED ISSUE #13446: amp-bind: Links break when mutating between hash and non-hash
@cramforce 

## What's the issue?

Clicking on the following link on an AMP page results in a broken AMP page when viewing in the Google AMP Viewer (only) in a Safari browser (only):

`<a on="tap:AMP.setState({showErrors: true})"
   class="searchButtonLink"
   [href]="searchForm.locationsUrlSegment && searchForm.dateUrlSegment && searchForm.paxNumber ? 'https://m.cheapoair.com/remotesearchhandler.aspx?t=f&tt=2' + searchForm.paxUrl + '&class='+searchForm.cabinClassUrlSegment+'&fpAffiliate=MobileFlightHomeDeals&from=' + searchForm.FromCityCode + '&to=' + searchForm.ToCityCode + '&fromDt=' + searchForm.departDate + '&toDt=' + searchForm.returnDate: '#'"
   href="#">
  <span class="ampstart-btn caps"
        id="search">Search</span>
</a>`

r04vatm77nh

This may be a bug in the cache rewrite in relation to amp-bind.

## How do we reproduce the issue?

Can no longer be produced as AMP pages have been taken down. 

## What browsers are affected?

All Safari browsers when viewing **from the Google AMP Viewer**

### PR DESCRIPTION
**Pull request by choumx**
Fixes #13446.

- When changing a link's href from non-hash to hash, remove `target` attribute (and store it)
- When changing a link's href from hash to non-hash, add `target=_top` (or the previously removed `target` value if applicable)

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -950,6 +938,30 @@ export class Bind {
...
+   * @param {string} value
+   * @return {string|undefined}
+   * @private
+   */
+  rewriteAttributes_(element, property, value) {
+    // Rewrite attributes if necessary. Not done in worker since it relies on
+    // `url#parseUrl` which uses <a>. Worker has URL API but not on IE11.
+    let rewrittenValue;
+    try {
+      rewrittenValue = rewriteAttributesForElement(element, property, value);
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: nit: just return here.

File: src/sanitizer.js [hunk 2]
```diff
@@ -410,6 +401,44 @@ export function isValidAttr(tagName, attrName, attrValue) {
...
+  const tag = element.tagName.toLowerCase();
+  const attr = attrName.toLowerCase();
+  const rewrittenValue = rewriteAttributeValue(tag, attr, attrValue);
+  // When served from proxy (CDN), changing an <a> tag from a hash link to a
+  // non-hash link requires updating `target` attribute per cache modification
+  // rules. @see amp-cache-modifications.md#url-rewrites
+  const isProxy = isProxyOrigin(opt_location || self.location);
+  if (isProxy && tag === 'a' && attr === 'href') {
+    const newValueIsHash = rewrittenValue[0] === '#';
+    const oldValueIsHash = element.getAttribute(attr)[0] === '#';
```
Comment (jridgewell) [inline src/sanitizer.js]: `getAttribute` can return null.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 1 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: I'd rather always have a `return` statement.

File: src/sanitizer.js (hunk 2 above)
Comment (choumx) [reply src/sanitizer.js]: Thanks!

**Pull request merged**